### PR TITLE
Add dim=None support to torch.logsumexp

### PR DESCRIPTION
**Pull request by stmcgovern**
Fixes   #156075

Update: Rebased and brought to date. 

In order to support dim=None for `torch.logsumexp`, I followed the pattern used by `torch.sum` and `torch.mean`, as far as possible. 

1. Changed the function signature in `native_functions.yaml` to make dimension argument optional
2. Changes in `ReduceOps.cpp` to modify (1) meta function and (2) impl function for logsumexp variants
3. Update `derivatives.yaml` and `FunctionsManual.c/h` for autograd support
4. Add explicit dim=None argument testing and implicit in test_reductions.py
5. Update documentation 

<del>
There were some changes made to dtype validation and type promotion handling that I following amin/amax for rather than sum and mean, since  the `opt_dtype` parameter is not present in the logsumexp signature.

These changes pass the subset of tests I ran on my cpu-only build `python -m pytest test/test_ops* -k "logsumexp" `, with the exception of 
(1) `test/test_ops.py::TestCommonCPU::test_meta_consistency_out_dtype_mismatch_logsumexp_cpu_float32` and 
(2) `test/test_ops_jit.py::TestJitCPU::test_variant_consistency_jit_logsumexp_cpu_float32`

From what I can tell, the mismatch issue is related to when the dtype consistency checks are made on the device path vs. meta path. The structured kernel dtype validation only runs on real devices, not on meta device. Meta device bypasses the structured kernel infrastructure that performs the dtype consistency checks. 

Running `python -m pytest test/test_reductions.py` passes the new test added. 

I would appreciate any guidance on the dtype consistency check mismatch. Thank you!
</del>

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @xmfan @H-Huang @chenyang78

### CODE REVIEW

**Comment by stmcgovern** [comment]
Upon further investigation, I conclude that `test_meta_consistency_out_dtype_mismatch_logsumexp_cpu_float32` should be expected to fail, just like `logcumsumexp` . I've also modified the schema for jit and updated test_reductions.py to reflect the new implementation. These changes make my local cpu build pass tests.

**Comment by stmcgovern** [comment]
Hi @albanD , I've updated the PR. I am seeing passing tests locally on CPU+CUDA. Please take a look when you can. Thanks!

**Comment by stmcgovern** [comment]
I'll look into the clang build failures first. I've been using gcc.

**Comment by stmcgovern** [comment]
thanks for the retest! I'll get a clang build next week and sort out the build failures. also address any failing tests.

**Comment by stmcgovern** [comment]
I'm just letting my clang build finish (which did hit the build error). I think it's as simple as removing a bad logsumexp_out function declaration, but I wanted to test locally. will follow up shortly after cleaning up.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3786,17 +3786,25 @@
...
+  dispatch:
+    CompositeExplicitAutograd: logsumexp
+
+- func: logsumexp.IntList_out(Tensor self, int[1]? dim, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True
   device_check: NoCheck   # TensorIterator
   dispatch:
-    # calls squeeze
-    CompositeExplicitAutogradNonFunctional: logsumexp_out
+    CPU, CUDA: logsumexp_out
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:3824]: You should keep this one CompositeExplicitAutogradNonFunctional to avoid breaking non-cpu/cuda backend right?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3786,17 +3786,25 @@
 
 - func: logcumsumexp.dimname_out(Tensor self, Dimname dim, *, Tensor(a!) out) -> Tensor(a!)
 
-- func: logsumexp(Tensor self, int[1] dim, bool keepdim=False) -> Tensor
+- func: logsumexp(Tensor self) -> Tensor
   device_check: NoCheck   # TensorIterator
   variants: function, method
   dispatch:
     CompositeExplicitAutograd: logsumexp
+  autogen: logsumexp.out
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: Why autogen this one rather than using the existing kernel?
The autogen is very inneficient

File: test/test_ops.py [hunk 3]
```diff
@@ -159,6 +159,7 @@ def reduction_dtype_filter(op):
     xfail("linalg.solve_ex"),
     xfail("linalg.solve_triangular"),
     xfail("logcumsumexp"),
+    xfail("logsumexp"),
```
Comment (albanD) [inline test/test_ops.py]: This is a new failure?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -873,8 +873,9 @@ Tensor logsumexp_backward(
     Tensor grad,
     const Tensor& self,
     Tensor result,
-    IntArrayRef dim,
+    OptionalIntArrayRef opt_dim,
     bool keepdim) {
+  auto dim = opt_dim.has_value() ? opt_dim.value() : IntArrayRef{};
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp]: Shouldn't this be the full range when no dim is passed as it is for a full reduction?

**Review by albanD — COMMENTED** [review]
Amazing work!
Thanks for going through all the different systems and updating them!

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (stmcgovern) [reply aten/src/ATen/native/native_functions.yaml:3824]: I wanted to try to keep it structured, so added a manual composite fallback. similar to all.dims_out and any.dims_out.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (stmcgovern) [reply aten/src/ATen/native/native_functions.yaml]: removed

File: test/test_ops.py (hunk 3 above)
Comment (stmcgovern) [reply test/test_ops.py]: It is similar to logcumsumexp (already on that list). The meta is more lenient than the actual composite ops. at::sum_out is invoked and it requires exact dtype match.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 4 above)
Comment (stmcgovern) [reply torch/csrc/autograd/FunctionsManual.cpp]: yes! thanks for catching this bug. the existing test passes since there is an implicit broacast. Added a test to verify intermediate shapes.

**Review by albanD — COMMENTED** [review]
Thanks!
The last thing would be to add examples for the new behavior in https://github.com/pytorch/pytorch/blob/fb64da07919e43f07c0a0f529c9848a95ad2d918/torch/testing/_internal/common_methods_invocations.py#L1568-L1599 so that the rest of the generic testing will also trigger on those.

**Review by albanD — APPROVED** [review]
Looks great!
Let's see what CI does!

**Comment by stmcgovern** [comment]
For the last failing test `test_modules_can_be_imported`, I added 2 missing overloads to NamedRegistries.cpp.

**Comment by stmcgovern** [comment]
@albanD I think my latest commit should address the failing modules test. Hopefully one more CI run will show green.

**Comment by albanD** [comment]
@pytorchbot merge

Trying to merge in case CI is fully green now!

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by stmcgovern** [comment]
not a priority at the moment.
